### PR TITLE
GLTF: Move accessor decoding functions to GLTFAccessor

### PR DESCRIPTION
**Pull request by aaronfranke**
Draft because this depends on #108853.

This PR moves the logic of the accessor decoding functions currently found in GLTFDocument into the GLTFAccessor class. The goal here is to allow for these functions to be used outside of GLTFDocument, such as by extensions, so that extensions can read and write accessor data in a type-safe way.

This is a prerequisite PR to allow for MultiMeshInstance3D import via EXT_mesh_gpu_instancing as seen in PR #107866.

Now, I said "move", but this PR is far from a trivial move. This is a refactor where I combed over the code multiple times to ensure that we actually have a good API free from bugs. Of course, it's possible that I've missed something, so this should be ideally be tested with importing many models before merging.

Here's a list of changes:

- Accessor decoding logic is now in GLTFAccessor instead of GLTFDocument. GLTFDocument has ~600 fewer lines of code in it, GLTFAccessor gains ~600 lines. Unlike the encode functions (see #108853), the existing decode logic wasn't so bad with code duplication, most of the logic was contained in just two functions.
- GLTFDocument retains wrapper functions to check if the accessor index is valid, get the accessor, call the GLTFAccessor functions, since the GLTFAccessor functions are instance methods. These wrappers retain the ability to optionally read indexed data as non-indexed data (used by meshes).
- `GLTFDocument::_decode_buffer_view` (really accessor logic) replaced with `GLTFAccessor::_decode_raw_numbers`
- `GLTFDocument::_decode_accessor` replaced with `GLTFAccessor::_decode_as_numbers`
- Both of those 2 functions are generic to allow for directly decoding numbers of the requested size, instead of always double. This fixes the ability to use 64-bit integer accessor types in extensions, and is potentially faster because it doesn't need to convert to a different size afterwards (probably the same speed though).
- 16-bit half-precision floats are now supported.
- Decoding sparse indices now has its own dedicated simpler function.
- This line `stride += 4 - (stride % 4); //according to spec must be multiple of 4` was bogus in the decoding logic, for the same reasons as it was in the encoding logic. Yes, the spec says it must be a multiple of 4... but if the files has the wrong stride, we can't just pretend it has a different one. Instead of "correcting", we need to error if this happens. With this PR, such invalid cases will print `"glTF import: The declared buffer view byte stride " + itos(declared_byte_stride) + " was not a multiple of 4 as required by glTF."`. Removing this also means we can drop the `p_for_vertex` parameter.

I tested this PR on real 3D models and they work correctly, and also stress-tested it on [these test files](https://github.com/KhronosGroup/glTF-Asset-Generator/tree/main/Output/Positive) which are specifically designed to contain all the edge cases of glTF and it works, but more testing is appreciated.

### CODE REVIEW

**Comment by fire** [comment]
I think we should merge this for 4.6, but what is the correct order of operations?

**Comment by aaronfranke** [comment]
@fire On every one of my draft PRs, the very top of the description has links to all the dependency PRs.

**Comment by fire** [comment]
It's blocked on https://github.com/godotengine/godot/pull/108853

File: modules/gltf/structures/gltf_accessor.cpp [hunk 1]
```diff
@@ -443,6 +443,63 @@ int64_t GLTFAccessor::_get_vector_size() const {
...
+		case Variant::RECT2:
+		case Variant::RECT2I:
+		case Variant::VECTOR4:
+		case Variant::VECTOR4I:
+		case Variant::PLANE:
+		case Variant::QUATERNION:
+		case Variant::COLOR:
+			return 4;
+		case Variant::TRANSFORM2D:
+		case Variant::AABB:
```
Comment (fire) [inline modules/gltf/structures/gltf_accessor.cpp:487]: Are you sure these are correct?

Comment (aaronfranke) [reply modules/gltf/structures/gltf_accessor.cpp:487]: glTF has no 2x3 matrix type, nor a 6D vector type, so for encoding a Transform2D semantically in glTF, the closest equivalent glTF type is the 3x3 matrix `"type": "MAT3"`, which is 9 floats. As for AABB, it similarly has 6 floats. Technically both of these could be encoded as 6 scalars, which would take up less space, but my assumption is that we want to leverage the existing type semantics in glTF, and glTF often requires padding accessor data. Note that I don't have a use case for AABB, my intention here was to support all Godot Variant types we could reasonably support.

**Review by fire — APPROVED** [review]

File: modules/gltf/structures/gltf_accessor.cpp [hunk 2]
```diff
@@ -443,6 +443,63 @@ int64_t GLTFAccessor::_get_vector_size() const {
 	ERR_FAIL_V(0);
 }
 
+int64_t GLTFAccessor::_get_numbers_per_variant_for_gltf(const Variant::Type p_variant_type) {
```
Comment (Repiteo) [inline modules/gltf/structures/gltf_accessor.cpp]:
```suggestion
int64_t GLTFAccessor::_get_numbers_per_variant_for_gltf(Variant::Type p_variant_type) {
```

File: modules/gltf/structures/gltf_accessor.cpp [hunk 3]
```diff
@@ -491,6 +548,546 @@ bool GLTFAccessor::is_equal_exact(const Ref<GLTFAccessor> &p_other) const {
...
+	ERR_FAIL_COND_V_MSG(numbers.size() != count * 4, ret, "glTF import: The accessor does not have the expected amount of numbers for the given count and vector size.");
+	ret.resize(count);
+	for (int64_t i = 0; i < count; i++) {
+		const int64_t number_index = i * 4;
+		ret.set(i, Quaternion(numbers[number_index], numbers[number_index + 1], numbers[number_index + 2], numbers[number_index + 3]).normalized());
+	}
+	return ret;
+}
+
+Array GLTFAccessor::decode_as_variants(const Ref<GLTFState> &p_gltf_state, const Variant::Type p_variant_type) const {
```
Comment (Repiteo) [inline modules/gltf/structures/gltf_accessor.cpp]:
```suggestion
Array GLTFAccessor::decode_as_variants(const Ref<GLTFState> &p_gltf_state, Variant::Type p_variant_type) const {
```

File: modules/gltf/structures/gltf_accessor.h [hunk 4]
```diff
@@ -90,9 +90,17 @@ class GLTFAccessor : public Resource {
 	static GLTFAccessorType _get_accessor_type_from_str(const String &p_string);
 	String _get_accessor_type_name() const;
 	int64_t _get_vector_size() const;
+	static int64_t _get_numbers_per_variant_for_gltf(const Variant::Type p_variant_type);
```
Comment (Repiteo) [inline modules/gltf/structures/gltf_accessor.h]:
```suggestion
	static int64_t _get_numbers_per_variant_for_gltf(Variant::Type p_variant_type);
```

File: modules/gltf/structures/gltf_accessor.h [hunk 5]
```diff
@@ -171,6 +179,18 @@ class GLTFAccessor : public Resource {
...
+	PackedColorArray decode_as_colors(const Ref<GLTFState> &p_gltf_state) const;
+	PackedFloat32Array decode_as_float32s(const Ref<GLTFState> &p_gltf_state) const;
+	PackedFloat64Array decode_as_float64s(const Ref<GLTFState> &p_gltf_state) const;
+	PackedInt32Array decode_as_int32s(const Ref<GLTFState> &p_gltf_state) const;
+	PackedInt64Array decode_as_int64s(const Ref<GLTFState> &p_gltf_state) const;
+	Vector<Quaternion> decode_as_quaternions(const Ref<GLTFState> &p_gltf_state) const;
+	PackedVector2Array decode_as_vector2s(const Ref<GLTFState> &p_gltf_state) const;
+	PackedVector3Array decode_as_vector3s(const Ref<GLTFState> &p_gltf_state) const;
+	PackedVector4Array decode_as_vector4s(const Ref<GLTFState> &p_gltf_state) const;
+	Array decode_as_variants(const Ref<GLTFState> &p_gltf_state, const Variant::Type p_variant_type) const;
```
Comment (Repiteo) [inline modules/gltf/structures/gltf_accessor.h]:
```suggestion
	Array decode_as_variants(const Ref<GLTFState> &p_gltf_state, Variant::Type p_variant_type) const;
```

File: modules/gltf/gltf_document.cpp [hunk 6]
```diff
@@ -954,647 +924,94 @@ Error GLTFDocument::_parse_accessors(Ref<GLTFState> p_state) {
...
-				}
-				ret.write[i] = p;
-			} break;
-			default: {
-				ERR_FAIL_V_MSG(ret, "glTF: Cannot decode accessor as Variant of type " + Variant::get_type_name(p_variant_type) + ".");
-			}
-		}
-	}
-	return ret;
+Array GLTFDocument::_decode_accessor_as_variants(const Ref<GLTFState> p_gltf_state, const GLTFAccessorIndex p_accessor_index, const Variant::Type p_variant_type) {
```
Comment (Repiteo) [inline modules/gltf/gltf_document.cpp]:
```suggestion
Array GLTFDocument::_decode_accessor_as_variants(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index, Variant::Type p_variant_type) {
```

File: modules/gltf/gltf_document.h [hunk 7]
```diff
@@ -133,53 +130,16 @@ class GLTFDocument : public Resource {
...
+	PackedInt32Array _decode_accessor_as_int32s(const Ref<GLTFState> p_gltf_state, const GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids = PackedInt32Array());
+	PackedVector2Array _decode_accessor_as_vec2(const Ref<GLTFState> p_gltf_state, const GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids = PackedInt32Array());
+	PackedVector3Array _decode_accessor_as_vec3(const Ref<GLTFState> p_gltf_state, const GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids = PackedInt32Array());
+	PackedColorArray _decode_accessor_as_color(const Ref<GLTFState> p_gltf_state, const GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids = PackedInt32Array());
+	Vector<Quaternion> _decode_accessor_as_quaternion(const Ref<GLTFState> p_gltf_state, const GLTFAccessorIndex p_accessor_index);
+	Array _decode_accessor_as_variants(const Ref<GLTFState> p_gltf_state, const GLTFAccessorIndex p_accessor_index, const Variant::Type p_variant_type);
```
Comment (Repiteo) [inline modules/gltf/gltf_document.h]:
```suggestion
	PackedFloat32Array _decode_accessor_as_float32s(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids = PackedInt32Array());
	PackedFloat64Array _decode_accessor_as_float64s(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids = PackedInt32Array());
	PackedInt32Array _decode_accessor_as_int32s(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids = PackedInt32Array());
	PackedVector2Array _decode_accessor_as_vec2(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids = PackedInt32Array());
	PackedVector3Array _decode_accessor_as_vec3(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids = PackedInt32Array());
	PackedColorArray _decode_accessor_as_color(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids = PackedInt32Array());
	Vector<Quaternion> _decode_accessor_as_quaternion(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index);
	Array _decode_accessor_as_variants(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index, Variant::Type p_variant_type);
```

File: modules/gltf/structures/gltf_accessor.cpp [hunk 8]
```diff
@@ -491,6 +548,546 @@ bool GLTFAccessor::is_equal_exact(const Ref<GLTFAccessor> &p_other) const {
...
+		}
+		numbers.set(i, number);
+		raw_read_offset += bytes_per_component;
+	}
+	ERR_FAIL_COND_V_MSG(raw_read_offset != raw_bytes.size(), numbers, "glTF import: Sparse indices buffer view size did not exactly match the expected size.");
+	return numbers;
+}
+
+template <typename T>
+Vector<T> GLTFAccessor::_decode_raw_numbers(const Ref<GLTFState> &p_gltf_state, const TypedArray<GLTFBufferView> &p_buffer_views, const bool p_sparse_values) const {
```
Comment (Repiteo) [inline modules/gltf/structures/gltf_accessor.cpp]:
```suggestion
Vector<T> GLTFAccessor::_decode_raw_numbers(const Ref<GLTFState> &p_gltf_state, const TypedArray<GLTFBufferView> &p_buffer_views, bool p_sparse_values) const {
```

File: modules/gltf/structures/gltf_accessor.h [hunk 9]
```diff
@@ -90,9 +90,17 @@ class GLTFAccessor : public Resource {
...
 	String _get_accessor_type_name() const;
 	int64_t _get_vector_size() const;
+	static int64_t _get_numbers_per_variant_for_gltf(const Variant::Type p_variant_type);
 	static int64_t _get_bytes_per_component(const GLTFComponentType p_component_type);
 	int64_t _get_bytes_per_vector() const;
 
+	// Private decode functions.
+	PackedInt64Array _decode_sparse_indices(const Ref<GLTFState> &p_gltf_state, const TypedArray<GLTFBufferView> &p_buffer_views) const;
+	template <typename T>
+	Vector<T> _decode_raw_numbers(const Ref<GLTFState> &p_gltf_state, const TypedArray<GLTFBufferView> &p_buffer_views, const bool p_sparse_values) const;
```
Comment (Repiteo) [inline modules/gltf/structures/gltf_accessor.h]:
```suggestion
	Vector<T> _decode_raw_numbers(const Ref<GLTFState> &p_gltf_state, const TypedArray<GLTFBufferView> &p_buffer_views, bool p_sparse_values) const;
```

File: modules/gltf/gltf_document.cpp [hunk 10]
```diff
@@ -954,647 +924,94 @@ Error GLTFDocument::_parse_accessors(Ref<GLTFState> p_state) {
...
-		ret_size = p_packed_vertex_ids.size();
-	}
-	ret.resize(ret_size);
-	for (int64_t i = 0; i < ret_size; i++) {
-		int64_t src_i = i;
-		if (!p_packed_vertex_ids.is_empty()) {
-			src_i = p_packed_vertex_ids[i];
-		}
-		ret.write[i] = Vector2(attribs_ptr[src_i * 2 + 0], attribs_ptr[src_i * 2 + 1]);
+PackedFloat32Array GLTFDocument::_decode_accessor_as_float32s(const Ref<GLTFState> p_gltf_state, const GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids) {
```
Comment (Repiteo) [inline modules/gltf/gltf_document.cpp]:
```suggestion
PackedFloat32Array GLTFDocument::_decode_accessor_as_float32s(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids) {
```

File: modules/gltf/gltf_document.cpp [hunk 11]
```diff
@@ -954,647 +924,94 @@ Error GLTFDocument::_parse_accessors(Ref<GLTFState> p_state) {
...
+	return _decode_unpack_indexed_data<PackedFloat32Array>(numbers, p_packed_vertex_ids);
 }
 
-Vector<Vector3> GLTFDocument::_decode_accessor_as_vec3(Ref<GLTFState> p_state, const GLTFAccessorIndex p_accessor, const bool p_for_vertex, const Vector<int> &p_packed_vertex_ids) {
-	const Vector<double> attribs = _decode_accessor(p_state, p_accessor, p_for_vertex);
-	Vector<Vector3> ret;
-
-	if (attribs.is_empty()) {
-		return ret;
+PackedFloat64Array GLTFDocument::_decode_accessor_as_float64s(const Ref<GLTFState> p_gltf_state, const GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids) {
```
Comment (Repiteo) [inline modules/gltf/gltf_document.cpp]:
```suggestion
PackedFloat64Array GLTFDocument::_decode_accessor_as_float64s(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids) {
```

File: modules/gltf/gltf_document.cpp [hunk 12]
```diff
@@ -954,647 +924,94 @@ Error GLTFDocument::_parse_accessors(Ref<GLTFState> p_state) {
...
-		ret_size = p_packed_vertex_ids.size();
-	}
-	ret.resize(ret_size);
-	for (int64_t i = 0; i < ret_size; i++) {
-		int64_t src_i = i;
-		if (!p_packed_vertex_ids.is_empty()) {
-			src_i = p_packed_vertex_ids[i];
-		}
-		ret.write[i] = Color(attribs_ptr[src_i * vec_len + 0], attribs_ptr[src_i * vec_len + 1], attribs_ptr[src_i * vec_len + 2], vec_len == 4 ? attribs_ptr[src_i * 4 + 3] : 1.0);
+PackedInt32Array GLTFDocument::_decode_accessor_as_int32s(const Ref<GLTFState> p_gltf_state, const GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids) {
```
Comment (Repiteo) [inline modules/gltf/gltf_document.cpp]:
```suggestion
PackedInt32Array GLTFDocument::_decode_accessor_as_int32s(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids) {
```

File: modules/gltf/gltf_document.cpp [hunk 13]
```diff
@@ -954,647 +924,94 @@ Error GLTFDocument::_parse_accessors(Ref<GLTFState> p_state) {
...
 
-	ERR_FAIL_COND_V(attribs.size() % 4 != 0, ret);
-	const double *attribs_ptr = attribs.ptr();
-	const int64_t ret_size = attribs.size() / 4;
-	ret.resize(ret_size);
-	{
-		for (int64_t i = 0; i < ret_size; i++) {
-			ret.write[i] = Quaternion(attribs_ptr[i * 4 + 0], attribs_ptr[i * 4 + 1], attribs_ptr[i * 4 + 2], attribs_ptr[i * 4 + 3]).normalized();
-		}
+PackedVector2Array GLTFDocument::_decode_accessor_as_vec2(const Ref<GLTFState> p_gltf_state, const GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids) {
```
Comment (Repiteo) [inline modules/gltf/gltf_document.cpp]:
```suggestion
PackedVector2Array GLTFDocument::_decode_accessor_as_vec2(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids) {
```

File: modules/gltf/gltf_document.cpp [hunk 14]
```diff
@@ -954,647 +924,94 @@ Error GLTFDocument::_parse_accessors(Ref<GLTFState> p_state) {
...
-	return ret;
+	return _decode_unpack_indexed_data<PackedVector2Array>(vectors, p_packed_vertex_ids);
 }
-Vector<Transform2D> GLTFDocument::_decode_accessor_as_xform2d(Ref<GLTFState> p_state, const GLTFAccessorIndex p_accessor, const bool p_for_vertex) {
-	const Vector<double> attribs = _decode_accessor(p_state, p_accessor, p_for_vertex);
-	Vector<Transform2D> ret;
 
-	if (attribs.is_empty()) {
-		return ret;
+PackedVector3Array GLTFDocument::_decode_accessor_as_vec3(const Ref<GLTFState> p_gltf_state, const GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids) {
```
Comment (Repiteo) [inline modules/gltf/gltf_document.cpp]:
```suggestion
PackedVector3Array GLTFDocument::_decode_accessor_as_vec3(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids) {
```

File: modules/gltf/gltf_document.cpp [hunk 15]
```diff
@@ -954,647 +924,94 @@ Error GLTFDocument::_parse_accessors(Ref<GLTFState> p_state) {
...
-		return ret;
-	}
-
-	ERR_FAIL_COND_V(attribs.size() % 9 != 0, ret);
-	ret.resize(attribs.size() / 9);
-	for (int64_t i = 0; i < ret.size(); i++) {
-		ret.write[i].set_column(0, Vector3(attribs[i * 9 + 0], attribs[i * 9 + 1], attribs[i * 9 + 2]));
-		ret.write[i].set_column(1, Vector3(attribs[i * 9 + 3], attribs[i * 9 + 4], attribs[i * 9 + 5]));
-		ret.write[i].set_column(2, Vector3(attribs[i * 9 + 6], attribs[i * 9 + 7], attribs[i * 9 + 8]));
+PackedColorArray GLTFDocument::_decode_accessor_as_color(const Ref<GLTFState> p_gltf_state, const GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids) {
```
Comment (Repiteo) [inline modules/gltf/gltf_document.cpp]:
```suggestion
PackedColorArray GLTFDocument::_decode_accessor_as_color(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index, const PackedInt32Array &p_packed_vertex_ids) {
```

File: modules/gltf/gltf_document.cpp [hunk 16]
```diff
@@ -954,647 +924,94 @@ Error GLTFDocument::_parse_accessors(Ref<GLTFState> p_state) {
...
-	ERR_FAIL_COND_V(attribs.size() % 16 != 0, ret);
-	ret.resize(attribs.size() / 16);
-	for (int64_t i = 0; i < ret.size(); i++) {
-		ret.write[i].basis.set_column(0, Vector3(attribs[i * 16 + 0], attribs[i * 16 + 1], attribs[i * 16 + 2]));
-		ret.write[i].basis.set_column(1, Vector3(attribs[i * 16 + 4], attribs[i * 16 + 5], attribs[i * 16 + 6]));
-		ret.write[i].basis.set_column(2, Vector3(attribs[i * 16 + 8], attribs[i * 16 + 9], attribs[i * 16 + 10]));
-		ret.write[i].set_origin(Vector3(attribs[i * 16 + 12], attribs[i * 16 + 13], attribs[i * 16 + 14]));
-	}
-	return ret;
+Vector<Quaternion> GLTFDocument::_decode_accessor_as_quaternion(const Ref<GLTFState> p_gltf_state, const GLTFAccessorIndex p_accessor_index) {
```
Comment (Repiteo) [inline modules/gltf/gltf_document.cpp]:
```suggestion
Vector<Quaternion> GLTFDocument::_decode_accessor_as_quaternion(const Ref<GLTFState> p_gltf_state, GLTFAccessorIndex p_accessor_index) {
```

**Review by Repiteo — APPROVED** [review]

**Comment by Repiteo** [comment]
Thanks!

**Comment by fire** [comment]
@akien-mga bisected a [gltf import] 1x to 2x slowdown because of this pull request on master. I'll post the most current master commit https://github.com/godotengine/godot/commit/4601c07d86bed2a5a25ea4b32c4f0b933c03e18b but the git bisect was done earlier.

aaronfranke is checking now.

**Comment by aaronfranke** [comment]
For the record: The specific part of the code that regressed in performance was about 50 times slower. So it was catastrophically bad, apologies for my mistake. Anyway, it is now fixed as of PR #113172.
